# 예외처리(exception handling)

프로그램이 실행 중 어떤 원인에 의해서 오작동을 하거나 비정상적으로 종료되는 경우가 있다. 이러한 결과를 초래하는 원인을 `프로그램 에러` 또는 `오류`라고 한다.

이를 발생시점에 따라 다음과 같이 나눌 수 있다.

> `컴파일 에러`: 컴파일 시에 발생하는 에러<br>
> `런타임 에러`: 실행 시에 발생하는 에러<br>
> `논리적 에러`: 실행은 되지만, 의도와 다르게 동작하는 것

소스 파일을 컴파일하면 컴파일러가 소스코드(`*.java`)에 대해 오타나 잘못된 구문, 자료형 체크 등의 기본적인 검사를 수행하여 오류가 있는지를 알려준다. 컴파일러가 알려준 에러들을 모두 수정해서 컴파일을 성공적으로 마치고 나면, 클래스파일이 생성되고, 생성된 클래스파일을 실행할 수 있는 것이다.

하지만 컴파일을 에러없이 성공적으로 마쳤다 하여 프로그램의 실행 시에도 에러가 발생하지 않는 것은 아니다. 컴파일러가 소스코드의 기본적인 사항은 컴파일 시에 모두 걸러 줄 수는 있지만, 실행도중에 발생할 수 있는 잠재적인 오류까지 검사할 수 없기 때문에 컴파일이 잘 되었어도 실행 중에 에러에 의해 잘못된 결과를 얻거나 프로그램이 비정상적으로 종료될 수 있다. 이를 `런타임 에러`라 하며, 프로그램이 실행 중 동작을 멈춘 상태로 오래 지속되거나, 갑자기 프로그램이 멈추고 종료되는 등이 이에 해당한다.

런타임 에러를 방지하기 위해서는 프로그램의 실행도중 발생할 수 있는 모든 경우의 수를 고려하여 이에 대한 대비를 하는 것이 필요하다. 자바에서는 실행 시(`runtime`) 발생할 수 있는 프로그램 오류를 에러(`error`)와 예외(`exception`)으로 구분한다.

에러는 메모리 부족(`OutOfMemoryError`)이나 스택오버플로우(`StackOverFlowError`)와 같이 일단 발생하면 복구할 수 없는 심각한 오류이고, 예외는 발생하더라도 수습이 가능한 비교적 덜 심각한 부류이다.

에러가 발생하면, 프로그램의 비정상적인 종료를 막을 수 없지만 예외는 발생하더라도 프로그래머가 적절한 코드를 미리 작성해 놓는 것으로 방지가 가능하다.

### 예외 클래스의 계층구조

### 예외처리: try-catch문

###
